# Coursera: Data Analysis Tools – Assignment 1
For this assignment the task was to run ANOVA test for some variables in the dataset selected in the previous course.
As my variables of interest are all categorical, I had to come up with a newly found quantitative variable, because ANOVA is used for the `C -> Q` cases where the explanatory variable is categorical and the response variable is quantitative.

My choice was to look at the relationship of cannabis use and perceived health. Hense my hypotheses:
* H<sub>0</sub>: There is no relationship between cannabis use and health perception;
* H<sub>A</sub>: There is some relationship between cannabis use and health perception.

My sample will be:
* The respondents who reported using cannabis within last 12 months at least once (N=1586)

The first step was to convert the information about cannabis use to quantitative values and form the sample. This variable initially contains 11 types of values, including 99 for Unknown, and also NaN values for the cases where the question is not applicable.

In [15]:
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import statsmodels.stats.multicomp as multi

NESARC = r"C:\Users\USER\Documents\Courses\Coursera\dai\datasets\nesarc\nesarc_pds.csv"

# Loading data
data = pd.read_csv(NESARC, low_memory=False)


### Processing variable reflecting recent cannabis use

CANNABIS_USE_12M = 'S3BD5Q2C'  # 'HOW OFTEN USED CANNABIS IN THE LAST 12 MONTHS'

# Convert values to numeric
data[CANNABIS_USE_12M] = pd.to_numeric(data[CANNABIS_USE_12M], errors='coerce')

# Code reference for CANNABIS_USE_12M
CANNABIS_USE_CAT_MAP = {
    1: "Every day",
    2: "Nearly every day",
    3: "3 to 4 times a week",
    4: "1 to 2 times a week",
    5: "2 to 3 times a month",
    6: "Once a month",
    7: "7 to 11 times a year",
    8: "3 to 6 times a year",
    9: "2 times a year",
    10: "Once a year",
    99: "Unknown",
}

# Get counts for CANNABIS_USE_12M
print(data[CANNABIS_USE_12M].value_counts(sort=False, dropna=False).rename(CANNABIS_USE_CAT_MAP))

NaN                     41479
3 to 6 times a year       192
Every day                 227
1 to 2 times a week       206
Nearly every day          121
2 times a year            149
Once a year               134
Unknown                    23
Once a month              165
2 to 3 times a month      175
3 to 4 times a week       117
7 to 11 times a year      105
Name: S3BD5Q2C, dtype: int64


So the NaN values were recoded to be `11`, to keep their distinction, while the Unknowns were recoded to be `NaN`.

In [16]:
# Recode meaningful NaN (to 11, N/A) and 'Unknowns' (99 to NaN)
data[CANNABIS_USE_12M] = data[CANNABIS_USE_12M].replace(np.NaN, 11).replace(99, np.NaN)
print(data[CANNABIS_USE_12M].value_counts(sort=False, dropna=False))

 11.0    41479
 8.0       192
 1.0       227
 4.0       206
 2.0       121
 9.0       149
 10.0      134
 6.0       165
 5.0       175
 3.0       117
NaN         23
 7.0       105
Name: S3BD5Q2C, dtype: int64


Next, the values for the cases of use were replaced by values reflecting the approximate number of use cases per the past 12 months. These values are based on the descriptive meanings of these values (such as `Every day => 365 cases per year`; `Nearly every day => 365 * .9 = 330 cases per year`; `3 to 4 times a week => 52 * 3.5 = 182 cases per year`, etc.).

In [17]:
# Recode values so that they reflect approximately number of times of cannabis use per last year, store in new column

CANNABIS_USE_TIMES_MAP = {
    1: 365,
    2: 330,
    3: 182,
    4: 104,
    5: 75,
    6: 12,
    7: 9,
    8: 4.5,
    9: 2,
    10: 1,
    11: 0
}

def recode_values(row):
    return CANNABIS_USE_TIMES_MAP.get(row[CANNABIS_USE_12M])

CANNABIS_USE_QUANT = 'CANNABIS_USE_QUANT'

data[CANNABIS_USE_QUANT] = data.apply(lambda row: recode_values(row), axis=1)

print(data[CANNABIS_USE_QUANT].value_counts(sort=False, dropna=False))

 0.0      41479
 1.0        134
 2.0        149
 4.5        192
 9.0        105
 330.0      121
 75.0       175
 365.0      227
 12.0       165
NaN          23
 104.0      206
 182.0      117
Name: CANNABIS_USE_QUANT, dtype: int64


These values were next converted to numeric to be used as quantitative data. Based on the newly created quantitative variable `CANNABIS_USE_QUANT` a subset was created only for the cases when cannabis was used in the past 12 months.

In [18]:
# Convert CANNABIS_USE_QUANT to numeric
data[CANNABIS_USE_QUANT] = pd.to_numeric(data[CANNABIS_USE_QUANT], errors='coerce')

# Make a subset only those who used cannabis within past 12 months
subset_cannabis = data[data[CANNABIS_USE_QUANT] > 0].copy()

print(subset_cannabis[CANNABIS_USE_QUANT].value_counts(sort=False, dropna=False))

4.5      192
1.0      134
2.0      149
9.0      105
104.0    206
12.0     165
330.0    121
75.0     175
182.0    117
365.0    227
Name: CANNABIS_USE_QUANT, dtype: int64


The next step was to handle the explanatory variable HEALTH_PERCEPTION. This variable has 6 possible values, one of which is Unknown. As this variable was going to be used in its categorical capacity only, its coded values were replaced by their text meanings.

In [19]:
# Introduce health perception as categorical variable
HEALTH_PERCEPTION = 'S1Q16'  # 'SELF-PERCEIVED CURRENT HEALTH'

# Reference for the health variable
HEALTH_VALUES = {
    1: 'Excellent',
    2: 'Very good',
    3: 'Good',
    4: 'Fair',
    5: 'Poor',
    9: 'Unknown'
}

# Convert values to numeric
subset_cannabis[HEALTH_PERCEPTION] = pd.to_numeric(data[HEALTH_PERCEPTION], errors='coerce')

# Recode Unknowns
subset_cannabis[HEALTH_PERCEPTION] = subset_cannabis[HEALTH_PERCEPTION].replace(9, np.nan)

# Replace codes by meanings

def decode_values(row):
    return HEALTH_VALUES.get(row[HEALTH_PERCEPTION])

subset_cannabis[HEALTH_PERCEPTION] = subset_cannabis.apply(lambda row: decode_values(row), axis=1)

print(subset_cannabis[HEALTH_PERCEPTION].value_counts(sort=False, dropna=False))

NaN            5
Good         406
Excellent    413
Very good    538
Fair         170
Poor          59
Name: S1Q16, dtype: int64


Lastly, a new dataframe was created containing only the two variables of interest: `CANNABIS_USE_QUANT` and `HEALTH_PERCEPTION`. All NaN values were dropped.
The ANOVA test was run on these variables.

In [20]:
# Create dataframe for only 2 variables
subset_cann_health = subset_cannabis[[CANNABIS_USE_QUANT, HEALTH_PERCEPTION]].dropna()

# Use .ols method to get the F-statistic and associated p-value
formula_health_cannab = '{} ~ C({})'.format(CANNABIS_USE_QUANT, HEALTH_PERCEPTION)
model_health_cannab = smf.ols(formula=formula_health_cannab, data=subset_cann_health)
results_health_cannab = model_health_cannab.fit()
print (results_health_cannab.summary())

                            OLS Regression Results                            
Dep. Variable:     CANNABIS_USE_QUANT   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                  0.012
Method:                 Least Squares   F-statistic:                     5.699
Date:                Fri, 08 Feb 2019   Prob (F-statistic):           0.000149
Time:                        12:28:46   Log-Likelihood:                -10034.
No. Observations:                1586   AIC:                         2.008e+04
Df Residuals:                    1581   BIC:                         2.010e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                97.51

The result showed F-statistic = 5.699 and p-value = 0.000149. As p-value is less than the 5% theashold, we are supposed to reject H<sub>0</sub> (that the two variables are not related) and accept H<sub>A</sub> (that they are related).
Therefore an association was established, but in a very general way, because the explanatory variable in this case is multilevel. To establish more precisely, which particular cases really demonstrate such an association, a post hoc test (Tuckey HSD) was used.

In [21]:
# Run post test
post_test = multi.MultiComparison(subset_cann_health[CANNABIS_USE_QUANT], subset_cann_health[HEALTH_PERCEPTION])
result_post_test = post_test.tukeyhsd()
print(result_post_test.summary())

  Multiple Comparison of Means - Tukey HSD,FWER=0.05  
  group1    group2  meandiff   lower    upper   reject
------------------------------------------------------
Excellent    Fair   27.3291   -6.3929   61.051  False 
Excellent    Good    35.829    9.9659  61.6922   True 
Excellent    Poor   58.6404    7.1356  110.1452  True 
Excellent Very good  8.9303   -15.2801 33.1407  False 
   Fair      Good     8.5     -25.3067 42.3066  False 
   Fair      Poor   31.3114   -24.6058 87.2285  False 
   Fair   Very good -18.3988  -50.9584 14.1608  False 
   Good      Poor   22.8114   -28.7489 74.3717  False 
   Good   Very good -26.8988  -51.2269 -2.5706   True 
   Poor   Very good -49.7102 -100.4615  1.0412  False 
------------------------------------------------------


The test showed three particular categorical values for which significant difference shows up in results:
* Excellent and Good;
* Excellent and Poor;
* Good and Very good.